File: 2473-max-sum-of-a-pair-with-equal-sum-of-digits/2473-max-sum-of-a-pair-with-equal-sum-of-digits.py
class Solution:
    def maximumSum(self, nums: List[int]) -> int:

        ans = -1  # Initialize answer to -1 (if no valid pair exists)
        dig = [0] * 90  # Array to store the largest number for each digit sum

        for x in nums:
            digit_sum = self.get_digit(x)  # Compute sum of digits of x

            if dig[digit_sum] != 0:  # If a number with the same digit sum exists
                t_sum = dig[digit_sum] + x  # Compute potential max sum
                ans = max(ans, t_sum)  # Update global max sum

                dig[digit_sum] = max(dig[digit_sum], x)  # Store the larger number
            else:
                dig[digit_sum] = x  # Store the first occurrence of a digit sum

        return ans

    def get_digit(self, num: int) -> int:
        s = 0
        while num != 0:
            s += num % 10
            num //= 10
        return s

File: 2473-max-sum-of-a-pair-with-equal-sum-of-digits/2473-max-sum-of-a-pair-with-equal-sum-of-digits.java
class Solution {
    public int maximumSum(int[] nums) {
        //To store digitsum->MAX(number)
        HashMap<Integer, Integer> maxdig = new HashMap<>();
        int maxSum = -1;

        for (int num : nums) {
            int digitSum = sum(num);

            if (maxdig.containsKey(digitSum)) {//Already koi hai phle se
                maxSum = Math.max(maxSum, num + maxdig.get(digitSum));

                if (maxdig.get(digitSum) < num)//current num wla is Bigger than old
                    maxdig.put(digitSum, num);
            } else
                maxdig.put(digitSum, num);
        }

        return maxSum;
    }

    private int sum(int num) {
        int s = 0;
        while (num != 0) {
            s += num % 10;
            num /= 10;
        }
        return s;
    }
}
